File: baekjoon/rust/zoac_16719.rs
use std::io::{stdin, stdout, BufRead, BufReader, BufWriter, Write};

fn print_word(word: &Vec<String>, show: &mut Vec<bool>) {
    let mut ans: Vec<String> = Vec::new();
    for i in 0..word.len() {
        if show[i] {
            ans.push(word[i].to_string());
        }
    }
    let mut writer = BufWriter::new(stdout());
    writeln!(writer, "{}", ans.join("")).unwrap();
    writer.flush().unwrap();
}

fn find_minimum_letter(word: &Vec<String>, left: usize, right: usize, show_index: &mut Vec<bool>) {
    if left == right {
        return;
    }

    let word_iter = word[left..right].iter();
    let mut zipped_word = word_iter
        .zip((0..word.len()).collect::<Vec<_>>())
        .collect::<Vec<_>>();
    zipped_word.sort_by_key(|v| v.0);

    let index = zipped_word[0].1;
    show_index[index + left] = true;
    print_word(word, show_index);

    find_minimum_letter(word, left + index + 1, right, show_index);
    find_minimum_letter(word, left, left + index, show_index);
}

fn solution(word: Vec<String>) {
    let mut show_index = vec![false; word.len()];
    find_minimum_letter(&word, 0, word.len(), &mut show_index);
}

fn main() {
    let mut reader = BufReader::new(stdin());
    let mut buffer = String::new();

    reader.read_line(&mut buffer).unwrap();
    let word = buffer
        .trim()
        .as_bytes()
        .iter()
        .map(|x| (*x as char).to_string())
        .collect::<Vec<_>>();
    solution(word);
}

File: baekjoon/python/zoac_16719.py
# Title: 문제제목
# Link: https://www.acmicpc.net/problem/16719

import sys
from collections import defaultdict
from collections import deque


sys.setrecursionlimit(10 ** 6)


read_single_str = lambda: sys.stdin.readline().strip()

def get_word(word: list, show_list: list, x: int):
    ans = ''
    for i in range(len(word)):
        if show_list[i]:
            ans += word[i]
        if i == x:
            ans += word[i]
    return ans


def solution_bf(word: list):
    show_idx = [False for _ in range(len(word))]
    for _ in range(len(word)):
        word_list = []
        for i in range(len(word)):
            if not show_idx[i]:
                word_list.append((get_word(word, show_idx, i), i))
        sorted_words = sorted(word_list, key= lambda x: x[0])
        show_idx[sorted_words[0][1]] = True
        print(sorted_words[0][0])


def print_word(word: list, show: list):
    ans = ''
    for i in range(len(word)):
        if show[i]:
            ans += word[i]
    print(ans)


def find_minimum_letter(word: list, left: int, right: int, show_index: list):
    if left == right:
        return

    ziped_word = zip(word[left:right], [i for i in range(len(word))])
    sorted_word = sorted(ziped_word, key=lambda x: x[0])

    index = sorted_word[0][1]
    show_index[index + left] = True
    print_word(word, show_index)

    find_minimum_letter(word, left+index+1, right, show_index)
    find_minimum_letter(word, left, left+index, show_index)


def solution_recursion(word: list):
    show_index = [False for _ in word]
    find_minimum_letter(word, 0, len(word), show_index)


def main():
    word = read_single_str()
    solution_recursion(list(word))


if __name__ == '__main__':
    main()